src/gluonnlp/models/t5.py

```diff
@@ -0,0 +1,849 @@
+        self._num_heads = num_heads
+        self._inner_dim = num_heads * d_kv
+        self._dtype = dtype
+        assert layout in ['TN', 'NT'], \
+            'Invalid layout: {}. Only "TN" and "NT" are supported.'.format(layout)
+        self._layout = layout
+
+        self.relative_position_encoder = RelAttentionScoreCell(
+            query_units=self._inner_dim, 
+            num_heads=num_heads, 
+            method='t5', 
+            bidirectional=(not is_decoder), 
+            embed_initializer=Normal(d_model ** -0.5 * init_factor), 
+            layout='NTK' if layout == 'NT' else 'TNK', 
+            dtype=dtype
+        )
+        self.layers = nn.HybridSequential()
+        for _ in range(num_layers): 
+            self.layers.add(
+                T5Block(
+                    d_model=d_model, 
+                    d_kv=d_kv, 
+                    d_ff=d_ff, 
+                    is_decoder=is_decoder, 
+                    num_heads=num_heads, 
+                    dropout_prob=dropout_prob, 
+                    layer_norm_eps=layer_norm_eps, 
+                    activation=activation, 
+                    init_factor=init_factor, 
+                    layout=layout, 
+                    dtype=dtype
+                )
+            )
+        self.final_layer_norm = T5LayerNorm(
+            d_model=d_model, 
+            epsilon=layer_norm_eps, 
+            init_factor=init_factor, 
+            dtype=dtype
+        )
+        self.dropout = nn.Dropout(dropout_prob)
+
+    @property
+    def layout(self): 
+        return self._layout
+
+    @_assert_decoder_method
+    @property
+    def state_batch_axis(self): 
+        return list(layer.state_batch_axis for layer in self.layers)
+
+    @_assert_decoder_method
+    def init_states(self, batch_size, ctx, dtype='float32'): 
+        return list(layer.init_states(batch_size, ctx, dtype) for layer in self.layers)
+
+    @_assert_decoder_method
+    def incremental_decode(
+        hidden_states, 
+        past_key_value, 
+        mem_states, 
+        mem_valid_length
+    ): 
+        raise NotImplementedError
+
+    def _get_query_key_shape(self, hidden_states, mem_states, past_key_value): 
+        # NT: (B, L_seq, inner_dim); TN: (L_seq, B, inner_dim)
+        index = 1 if self.layout == 'NT' else 0
+        query_length = hidden_states.shape[index]
+        if past_key_value is not None: 
+            # for incremental decoding only, where past key and value are of shape
+            # NT(NTK): (B, L_seq, num_heads, n_kv); TN(TNK): (L_seq, B, num_heads, n_kv)
+            query_length += past_key_value[0].shape[index]
+        key_length = query_length if mem_states is None else mem_states.shape[index]
+        return query_length, key_length
+
+    @staticmethod
+    def _get_relative_position(query_length, key_length): 
+        # relative_position = mem_i - query_j
+        query_position = np.arange(query_length, dtype=np.int32)[:, None]
```

Comment:
For example, assume data has shape `(B, query_length, mem_length)`, then `npx.arange_like(hidden_states, axis=1)` will be `[0, 1, 2, ..., query_length - 1]`.

Reply:
I used `expand_dims` instead and now the hybridization works fine. Also, `hybridize` seems to support `.shape` now. Thank you for pointing it out though.